### PR TITLE
Add imaging

### LINKED ISSUE #1: [TODO]: Add Imaging data
The Imaging data is quite detailed.

### Note on data acquisition

The custom microscope controller initially writes directly to binary for speeds, but a secondary step writes it to the HDF5 files we observe, mainly to benefit from the automatic compression. Unfortunately, older versions of this secondary step had the potential to delete the binary files without writing the subsequent HDF5 files if there was not enough system memory to complete the HDF5 conversion step (binary files were deleted automatically after the HDF5 step 'completed', the mistake being it would not check for success or failure). Later versions corrected this but the result is the data stored on certain tapes might appear to us as was initially shared - with large portions of the original data completely missing.

### Prototype data

For the data initially shared, `20170228/fish4/20170228_4_1_gfaprgeco_hucgc_6dpf_shorttrials_20170228_185002`

There are 5777 `.h5` files in the `/raw/` folder and 285 `.tif` files in the `/projections/` folder. There should be, however, 7200 complete raw images and 7200 corresponding 'projections'. The fact that there are not is due to the issue from the first section.

Each `.h5` file contains a single 29 x 2048 x 2048 dataset. The first axis is the number of volumetric stacks, and the rest is the image size.

The 'projection' TIFF files are essentially summary images - they difference the fluorescence over the images, and then reduce each image into 4x4 bins as well as collapse the 29 stacks down to 8. However, it is simply translated from their pixel-wise registration method - there is no rotation as was done as a part of official segmentation. Data is largely used for visualization purposes.

The 'average' sampling frequency is `1.58` Hz as indicated in a number of locations. However, precise shutter opening/closing times are available from the sync channels in the `ephys` subfolder, and those should be used to get exact timestamps (sampling rate of sync systems was MUCH higher).

### Log files

There is a file called `Stack dimensions.log`, it can be read via

```python
with open(name="...", mode="rb") as file:
    values = list(file.read())
```

which gives a list of 3 non-zero values separated by a bunch of zeros. Those non-zero values are `8`, `8`, and `29`. The `29` makes sense for the volumetric stacking of the imaging planes, unsure about the others.

According to Yu this doesn't really matter at all.

There is another file `Stack_frequency.txt` that can be read as text and has the values and commented roles

```
1.58  # Average sampling rate in Hz
4558.46  # total time in seconds spanned by recording BUT see comment above on exact timestamps since they will be irregular
7200  # total number of frames
```

### Visualization (reference figure and rotational alignment)

There are also two files one level above the imaging subfolder called `anat_reference.tif` and `regparams_affine.npy`. 

`regparams_affine.npy` is a numpy-pickled `7200 x 3 x 3` array describing the affine transform to use to rotation each volume to align to the `anat_reference.tif` for visualization purposes.

### Note for visualizing source data

A cool thing I just found for visualizing the content of the .h5 files is HDFView supports images!

image

Just alter the read dimensions to the 2nd and 3rd axes and then you can click through the volume stacks.

### PR DESCRIPTION
**Pull request by CodyCBakerPhD**
Precedes #4. Introduces a bit of package structure as well for installation (this was cookie-cuttered from a way earlier version that didn't have everything in it yet)

### CODE REVIEW

File: src/ahrens_lab_to_nwb/yu_mu_cell_2019/yu_mu_cell_2019_imaging_extractor.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+    def get_video(self, start_frame: Optional[int] = None, end_frame: Optional[int] = None) -> np.ndarray:  # noqa: D102
+        with h5py.File(name=self.file_path) as file:
+            video = DatasetView(file["default"]).lazy_transpose(self._frame_axis_order)
+            return video.lazy_slice[:, :, :].dsetread()[np.newaxis, :]
+
+    def get_image_size(self) -> Tuple[int, int, int]:  # noqa: D102
+        return (self._num_rows, self._num_cols, self._num_stacks)
+
+    def get_num_frames(self):  # noqa: D102
```
Comment (h-mayorquin) [inline src/ahrens_lab_to_nwb/yu_mu_cell_2019/yu_mu_cell_2019_imaging_extractor.py:46]: How comes 1 ?

File: src/ahrens_lab_to_nwb/yu_mu_cell_2019/yu_mu_cell_2019_imaging_extractor.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            self._dtype = file["default"].dtype
+        self._frame_axis_order = [1, 2, 0]
+
+    def get_video(self, start_frame: Optional[int] = None, end_frame: Optional[int] = None) -> np.ndarray:  # noqa: D102
+        with h5py.File(name=self.file_path) as file:
+            video = DatasetView(file["default"]).lazy_transpose(self._frame_axis_order)
+            return video.lazy_slice[:, :, :].dsetread()[np.newaxis, :]
+
+    def get_image_size(self) -> Tuple[int, int, int]:  # noqa: D102
+        return (self._num_rows, self._num_cols, self._num_stacks)
```
Comment (h-mayorquin) [inline src/ahrens_lab_to_nwb/yu_mu_cell_2019/yu_mu_cell_2019_imaging_extractor.py:44]: `self.num_stacks`? I guess that's a good name for a tower of images. Did you see this somewhere?

**Review by h-mayorquin — COMMENTED** [review]
Everything looks OK to me. You define a specific ImagingExtractor to deal with the volumetric data as our HDF5 extractor does not have this flexibility (maybe it should). Then you concatenate the files using the MultiImagingExtractor by appropriately sorting the files.

I guess I am just curios about the metadata as I don't know anything about this type of Microscope. 
For the device?
Are they called Microscopes even if they can read volumetric data? Anything about the manufacturer?

For the ImagingPlane any type of reference of how the grid is done? Space between the stacks? What's the origin? Area of the brain (location) or imaging indicator that is used?

Anything about the Optical Channel that is actually recorded?

**Comment by CodyCBakerPhD** [comment]
> Are they called Microscopes even if they can read volumetric data? Anything about the manufacturer?

There's a little bit of info but it'll have to be added manually and I was saving that for a subsequent polishing step. I think we have a manufacturer and model number though, but the software that runs it is all custom.

Yeah, it's still a 'light-sheet micoscope'. The volumetric side comes from using multi-photon excitation that allows better differentiation against background fluorescence. Also the microscope is just what amplifies the image captured by the camera, not perhaps a reference to the entire rig design itself (which we've considered out of scope for NWB currently).

> For the ImagingPlane any type of reference of how the grid is done? Space between the stacks?


Slight overlap between stacks but not quantified. If there were space between stacks they would be written as difference planes. No information about quantitative distances spanned.

> What's the origin? Area of the brain (location)

I'll add a reference image in a future PR that will explain the origin - I was waiting on widget debugging/development to really make use of that.

It's whole brain though.

> or imaging indicator that is used? Anything about the Optical Channel that is actually recorded?

Calcium indicator also added manually with the excitation/emission info in the subsequent 'polishing metadata' PR. This PR is largely just the data engineering to get there.

File: src/ahrens_lab_to_nwb/yu_mu_cell_2019/yu_mu_cell_2019_imaging_extractor.py (hunk 2 above)
Comment (CodyCBakerPhD) [reply src/ahrens_lab_to_nwb/yu_mu_cell_2019/yu_mu_cell_2019_imaging_extractor.py:44]: Not sure where I first saw it or why I even chose that name at the time. They aren't separate planes so the nomenclature of stuff like the SbxImagingExtractor isn't exact...

https://nwb-schema.readthedocs.io/en/latest/format.html?highlight=stack#sec-correctedimagestack-src is the closest reference to that I found on NWB.

Maybe the word just came up in reading the paper and talking with the experimenter

File: src/ahrens_lab_to_nwb/yu_mu_cell_2019/yu_mu_cell_2019_imaging_extractor.py (hunk 1 above)
Comment (CodyCBakerPhD) [reply src/ahrens_lab_to_nwb/yu_mu_cell_2019/yu_mu_cell_2019_imaging_extractor.py:46]: From the issue with all the details: https://github.com/catalystneuro/ahrens-lab-to-nwb/issues/1

Each file (this extractor reads a single file from a big folder) is a single frame, hence the MultiImaging to get a true recording out of it

**Comment by h-mayorquin** [comment]
> Slight overlap between stacks but not quantified. If there were space between stacks they would be written as difference planes. No information about quantitative distances spanned.

Are you sure about the different planes? I was reading the schema and came out with the different impression. `TwoPhotonSeries` is only linked to one `ImagingPlane` even if has depth. And the ImaigngPlane used to contain a manifold and now contains origin coordinates and grid spacing which indicates geometrical structure.

File: src/ahrens_lab_to_nwb/yu_mu_cell_2019/yu_mu_cell_2019_imaging_extractor.py (hunk 1 above)
Comment (h-mayorquin) [reply src/ahrens_lab_to_nwb/yu_mu_cell_2019/yu_mu_cell_2019_imaging_extractor.py:46]: I see, that explains the np.newaxis as well and why you only reference three dimensions in the transposition per file. That clicks now.

**Review by h-mayorquin — APPROVED** [review]
Fair enough, everything here looks good to me from far away and the metadata discussion will come up later as Cody mentions. 

One thing that I am not sure is if the `get_video` method is working lazily as it produces a numpy array per frame that is concatenated with the `MultiImagingExtractor` but I guess that @CodyCBakerPhD tested this as the data seems to big to have worked otherwise. 

Given that you read only one frame per extractor and then you transform to numpy per output (call to `dsetread`) is it using lazyops actually useful here?

**Comment by h-mayorquin** [comment]
Btw, this was useful for me. I was under the impression that the PEP8 disavowed underscores in module names but they are actually flexible on this one:
https://peps.python.org/pep-0008/#package-and-module-names

I have some horrible module names in Murthy conversion. I will change this and probably the cookie cutter template generator.

**Comment by h-mayorquin** [comment]
Oh, yes, and what about the strange `sampling_frequency=1.56`?

**Comment by CodyCBakerPhD** [comment]
> Are you sure about the different planes?

Comes down to how the ROIs are defined. If ROIs are segmented per-plane then you want multiple NWB imaging planes. If the planes are close enough together that an ROI is segmented across multiple stacks, then a single NWB plane must be used for the ROI table reference from the RoiResponseSeries to link appropriately, which is the situation here.

> I was reading the schema and came out with the different impression.

I don't think the schema explains the difference very well. I also doubt there's a tutorial in the PyNWB docs that explains it well either given I had to write one just for using the `voxel_masks`.

> TwoPhotonSeries is only linked to one ImagingPlane even if has depth.

Right, you'd also do a different TwoPhotonSeries if you were scanning planes that were separated by a large enough distance.

> And the ImaigngPlane used to contain a manifold and now contains origin coordinates and grid spacing which indicates geometrical structure.

`manifold` is deprecated (https://nwb-schema.readthedocs.io/en/latest/format.html#imagingplane) but you're right about the grid spacing, I just have no information on that. The origin coords will be available when I attach a reference image, I'm not sure which corner/rotation it is right off the bat.

**Comment by CodyCBakerPhD** [comment]
> Oh, yes, and what about the strange sampling_frequency=1.56?

Nothing strange about it, that's the sampling speed for a total volume. It is 'specified' loosely in a couple files but I'm hard-coding it in the script for now since those file fields are largely hard-coded as well.

In a follow-up PR I'll also be synchronizing it directly with the shutter tracking system in the behavior, which is much higher resolution (thousands of Hz) and tracks the exact timestamps for the imaging data, so that rate won't be used much or possibly at all in the end.

**Comment by CodyCBakerPhD** [comment]
> Btw, this was useful for me. I was under the impression that the PEP8 disavowed underscores in module names but they are actually flexible on this one:

Right, just not in packages, in which case the underscores are swapped for dashes I think. But then you import the package with an underscore where there was a dash?

**Comment by CodyCBakerPhD** [comment]
> One thing that I am not sure is if the get_video method is working lazily as it produces a numpy array per frame that is concatenated with the MultiImagingExtractor but I guess that @CodyCBakerPhD tested this as the data seems to big to have worked otherwise.
> 
> Given that you read only one frame per extractor and then you transform to numpy per output (call to dsetread) is it using lazyops actually useful here?

The `MultiImagingExtractor` call to `get_video` deploys the calls to sub-extractors on-demand per span of slice requested, so the data on a per-extractor basis is indeed loaded into memory and returned at that point. It would be great if a memory map could be maintained as long as possible but I couldn't figure out to do it with the new axis.

From a point of view of data chunk iteration for the final `io.write()` call you're only skipping a couple steps before it gets loaded into memory anyhow. But if you wanted to use it for general purpose data retrieval yeah you would need to be careful

Yeeeah I suppose since the laziness is broken pretty quickly there anyway the `lazy_ops` isn't necessary unless it's faster to transpose it on-read as opposed to in-memory for some reason (I would hope not though). I'll test removing that in a follow up, thanks for pointing that out.

**Pull request merged**